collector-sqs/src/main/java/zipkin/collector/sqs/SQSSpanProcessor.java

```diff
@@ -61,59 +61,72 @@
 
   @Override public void close() throws IOException {
     if (closed.getAndSet(true)) throw new IllegalStateException("SQS processor is already closed.");
+    runner.getAndUpdate((t) -> {t.interrupt(); return t;});
     client.shutdown();
   }
 
   SQSSpanProcessor run() {
-    // don't throw an exception here since this might be run from a receive callback.
-    if (closed.get()) return null;
-
-    ReceiveMessageRequest request = new ReceiveMessageRequest(queueUrl)
-        .withWaitTimeSeconds(waitTimeSeconds);
-
-    client.receiveMessageAsync(request,
-        new AsyncHandler<ReceiveMessageRequest, ReceiveMessageResult>() {
-          @Override public void onError(Exception exception) {
-            if (exception instanceof AmazonClientException) {
-              // TODO should we ever give up if its retryable?
-              if (((AmazonClientException)exception).isRetryable()) {
-                run();
-              }
-            } else {
-              logger.log(Level.WARNING, "receive failed", exception);
-              status.set(CheckResult.failed(exception));
-            }
-          }
+    if (closed.get()) {
+      throw new IllegalStateException("Can not call run on a closed processor");
+    }
+    if (runner.get() != null) {
+      throw new IllegalStateException("Processor is already running");
+    }
 
-          @Override
-          public void onSuccess(ReceiveMessageRequest request, ReceiveMessageResult result) {
-            process(result.getMessages());
-            status.lazySet(CheckResult.OK);
-            run();
-          }
-        });
+    final ReceiveMessageRequest request = new ReceiveMessageRequest(queueUrl)
+        .withWaitTimeSeconds(waitTimeSeconds)
+        .withMaxNumberOfMessages(maxNumberOfMessages);
+
+    runner.updateAndGet(t -> run(t, request));
```

Comment:
The run method isn't meant to be re-entrant.  The atomic variable was meant to avoid race between run/close.  In hind sight this is confusing and I should probably just use synchronize blocks.